ui/src/components/build/CreateBuildConfigDialog.js

```diff
@@ -0,0 +1,189 @@
+      KEY_CR_BUILD_IOS_CREDENTIALS,
+      KEY_CR_BUILD_IOS_CREDENTIALS_DEVELOPER_PROFILE,
+      KEY_CR_BUILD_IOS_CREDENTIALS_NAME
+    ]);
+  }
+
+  componentDidUpdate(prevProps) {
+    this.validate();
+    const { initialConfig: prevConfig = {} } = prevProps;
+    const { initialConfig = {} } = this.props;
+    if (!isEqual(prevConfig, initialConfig)) {
+      this.configState.set(KEY_CR_CLIENT_ID, initialConfig[KEY_CR_CLIENT_ID]);
+      this.configState.set(KEY_CR_CLIENT_TYPE, initialConfig[KEY_CR_CLIENT_TYPE]);
+      switch (initialConfig[KEY_CR_CLIENT_TYPE]) {
+        case CLIENT_TYPE_ANDROID:
+          this.buildState.set(KEY_CR_BUILD_PLATFORM, BUILD_PLATFORM_ANDROID);
+          this.setState({ [KEY_HIDE_PLATDORM]: true });
+          break;
+        case CLIENT_TYPE_IOS:
+          this.buildState.set(KEY_CR_BUILD_PLATFORM, BUILD_PLATFORM_IOS);
+          this.setState({ [KEY_HIDE_PLATDORM]: true });
+          break;
+        default:
+          this.setState({ [KEY_HIDE_PLATDORM]: false });
+      }
+    }
+  }
+
+  validate() {
+    let validVal = this.configState.isAllValid() && this.buildState.isAllValid() && this.sourceState.isAllValid();
+    if (this.sourceState.get(KEY_CR_SOURCE_AUTH_TYPE) === BUILD_AUTH_TYPE_BASIC)
+      validVal = validVal && this.basicAuthState.isAllValid();
+    if (this.sourceState.get(KEY_CR_SOURCE_AUTH_TYPE) === BUILD_AUTH_TYPE_SSH)
+      validVal = validVal && this.sshAuthState.isAllValid();
+    if (this.state.valid !== validVal) this.setState({ valid: validVal });
+  }
+
+  onSaveBuildConfig = () => {
+    const { config } = this.state;
+    const { onSave, update = false, initialConfig } = this.props;
+    onSave && onSave({ update, config: { ...initialConfig, ...config } });
+  };
+
+  cancel = () => {
+    const { onCancel } = this.props;
+    onCancel && onCancel();
+  };
+
+  renderFormButtons = (onCancel, update, valid) => (
+    <div>
+      <Button bsStyle="primary" onClick={this.onSaveBuildConfig} disabled={!valid}>
+        {update ? 'Save' : 'Create'}
+      </Button>
+      <Button onClick={onCancel}>Cancel</Button>
+    </div>
+  );
+
+  getValidationState = () => {};
+
+  render() {
+    const { show, title, onCancel, update } = this.props;
+    const { valid = false } = this.state;
+    return (
+      <Modal show={show} container={this}>
+        <Modal.Header>
+          <Modal.CloseButton onClick={onCancel} />
+          <Modal.Title>{title}</Modal.Title>
+        </Modal.Header>
+        <Modal.Body className="modalBody">
+          <Grid fluid>
+            <Row>
+              <Col md={12}>
+                <div className="help-block">
+                  Client Build allows you to create a mobile client binary from a git source repository.
+                </div>
+                <Form>
+                  {renderNameSection(this.configState)}
+                  {renderSourceSection(this)}
+                  {renderBuildSection(this)}
```

Comment:
IMO only parameters that should be passed to these render methods are relevant parts of the state and callbacks for handling changes.

Reply:
Redux solves this now.